### ISSUE TITLE
deploy.sh updates defaults files rather than add to overrides

### ISSUE THREAD

**Issue by mattt416** [rcbops/rpc-openstack#1436]
###### Please provide the output to 'git describe --all' for  all deployments where you have confirmed it occurs.

Mitaka versions of scripts/deploy.sh should override variables by adding them to the respective overrides files, not overwriting them in the defaults files.

##### What type of issue is this? Please tick the appropriate box.
- [x] bug - existing functionality that does not work for you as documented
- [ ] feature - new functionality that you would like to see added

##### What is the expected behavior?

Variable overrides should be added to the overrides files, not updated in the defaults files.

##### What is actually happening?

Variables are being updated in the defaults files.

##### What were you attempting to do when you encountered this error (give as much detail as possible)? 

No error, just a bad practice.

##### How can we reproduce this issue?

Deploy an AIO using scripts/deploy.sh.

##### How can we verify your issue is addressed?

The overrides files are correctly populated with variable overrides.

##### What is your evaluation on the importance of this issue, and why?

----------------------------------------------------------------------------------------------------------------------------
###### For each pull request, update the branch line to reference it.
- [x] master https://github.com/rcbops/rpc-openstack/pull/1444
- [ ] mitaka-13.1
- [ ] mitaka-13.0
- [ ] liberty-12.2
- [ ] kilo

##### Upgrade impact. Consider the impact on moving between major, minor and patch releases.
- [ ] This change does not impact upgrades
- [ ] This change addresses a potential upgrade impact
- [ ] The upgrade impact is being tracked by another issue.

The issue tracking the upgrade impact is:

##### Testing impact
- [ ] An issue/pull-request has been create or exists to track the testing gap, and is referenced below
- [ ] This change includes additional testing
- [ ] This change is tested by the existing setup

Additional testing links:

##### Release note
- [ ] This change includes a release note.
- [ ] This change does not require a release note.

##### Does your change require documentation?
- [ ] Yes, and I submitted an issue to [Docs issues](https://github.com/rackerlabs/docs-rpc/issues "Docs issues")
- [ ] Yes, and I contributed a documentation patch to [docs-rpc](https://github.com/rackerlabs/docs-rpc "docs-rpc")
- [ ] No documentation required for this change

Additional documentation links (please add links if you have ticked yes):

**Comment by git-harry** [comment]
@mattt416 would you mind adding examples of what you're referring to. Based on the description, it seems this is an AIO specific issue, I've marked this as a prio-3 but we can adjust if required based on any additional information.

**Comment by evrardjp** [comment]
@git-harry We discovered this when doing an AIO indeed.
In the deploy.sh, if AIO, we override the ceph variables in the defaults. 

It seems wrong, because what needs to be effectively set is the override, we are like a deployer that sets these variable. However, if we explain to support we never touch the override file, this is a problem.

If we omit this problem of expectations of "scoping", we still have to remember that these default variables are overriden by the _overrides (which makes sense, right?). 
During the test-upgrades.sh we are merging variables for the deployer. 
By default on my AIO, I ended up by having ``osd_directory: True`` and ``raw_multi_journal: True`` due to the merging, where (IMO) I should only have ``osd_directory: True`` (``raw_multi_journal: False``), because I set my env variables correctly.

To sum up, this is a bug due to expectations:
- I expected that setting my env variables will do proper overriding
- I expected that the upgrade merging of the variables wouldn't touch the overrides files (and therefore would go against the env variables set in the deploy.sh happening after).

**Comment by git-harry** [comment]
After discussing with @evrardjp this is blocking https://github.com/rcbops/u-suk-dev/issues/383 (r13.1.0rc2) so moving to approved.

**Comment by evrardjp** [comment]
Well my comment above is completely irrelevant. 
Please ignore it, and put the prio back to its original state.
We should remove the blocking part too.

**Comment by prometheanfire** [comment]
It was my understanding that overrides were for support...  This is working as intended as far as I can see.  @evrardjp so is this still an issue?

**Comment by alextricity25** [comment]
IMO, variables added/modified in the ``deploy.sh`` **should either be defined as the correct value in the defaults file already, or be added to the overrides file**. Right now it's adding/overriding directives in the defaults file(usually in the case of an AIO). I would rather see these directives in the overrides file.

**Comment by prometheanfire** [comment]
The deploy.sh script alters the default the question is if we want to consider those alterations as being done by dev or deployers?  If by dev then it should stay in defaults, if by deployers then it should be inserted into overrides.  I think there's a good argument for stating that choices made by deploy.sh are a deployer decision (deploy(er).sh) :D

**Comment by mattt416** [comment]
I started a [WIP](https://github.com/mattt416/rpc-openstack/commit/efda8ee1b96131487bb353c41e780792fe5057f6) with some ideas, but at the moment I'm not sure which direction to go.  The [WIP](https://github.com/mattt416/rpc-openstack/commit/efda8ee1b96131487bb353c41e780792fe5057f6) effectively does the following:

- creates a bash function called `update_overrides` so that the logic of updating the overrides only needs to exist in one place
- fixes all the overriding in `scripts/deploy.sh` to use the correct overrides files (not to update the defaults files)
- moves all the overrides into a separate file (with the exception of the security stuff, still need to do that), with the idea being that all the base RPC overrides can be re-set later
- updates `scripts/test-upgrade.sh` and `scripts/deploy.sh` to include the new file to set/reset overrides

Now, the issues:

1. If we attempt to update the value of an override in `scripts/test-upgrade.sh` (necessary after migrate-yaml.py has run), then that gets a bit tricky with bash when the value is not a simple string.  As such, I had to embed some nasty python into `scripts/functions.sh` so that we can simply delete the entry and then rewrite it.
2. Including `scripts/deploy-rpc-overrides.py` into `scripts/test-upgrade.sh` means we now need to set a bunch more variables in `scripts/test-upgarde.sh` which we didn't need to before.
3. It all leaves me with a rather bitter taste in my mouth.  :)

@alextricity25 @evrardjp @git-harry @prometheanfire would appreciate your feedback here before I actually create a PR for any of this work.

**Comment by evrardjp** [comment]
You're right we are creating a super complex thingy for something that shouldn't be.

Thinking of a better solution.

**Comment by alextricity25** [comment]
> I feel like we're adding some complexity for manually testing scripts/test-upgrade.sh on an AIO. This script is specifically for gate testing, and I do wonder if it'd not be better to just ensure that any var that gets set in scripts/deploy.sh has an appropriate override in jenkins' specific overrides file.

I agree with this ten fold! We shouldn't have to cater this script to everyone(support, gate, and devs). It's meant only for gate testing. Period. It would be nice to add a NOTE at the top explaining the intended use for the script.
``test-upgrades.sh`` can also be used by DEVs on a local environment, but only with the understanding that their might be some extra work to do so the script can run. In the ceph case, setting ``raw_multi_node: False`` in the overrides file.
